8313657 : com.sun.jndi.ldap.Connection.cleanup does not close connections on SocketTimeoutErrors

weibxiao opened the pull request:
com.sun.jndi.ldap.Connection::leanup does not close the underlying socket if the is an IOException generation when the output stream was flushing the buffer.

Please refer to the bug https://bugs.openjdk.org/browse/JDK-8313657.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313657](https://bugs.openjdk.org/browse/JDK-8313657): com.sun.jndi.ldap.Connection.cleanup does not close connections on SocketTimeoutErrors (**Bug** - P3)


### Reviewers
 * [Vyom Tewari](https://openjdk.org/census#vtewari) (@vyommani - Committer) ⚠️ Review applies to [e3e98226](https://git.openjdk.org/jdk/pull/15143/files/e3e98226a9771f2054eb203cd7ef1db322be1d7a)
 * [Mark Sheppard](https://openjdk.org/census#msheppar) (@msheppar - **Reviewer**)
 * [Aleksei Efimov](https://openjdk.org/census#aefimov) (@AlekseiEfimov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15143/head:pull/15143` \
`$ git checkout pull/15143`

Update a local copy of the PR: \
`$ git checkout pull/15143` \
`$ git pull https://git.openjdk.org/jdk.git pull/15143/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15143`

View PR using the GUI difftool: \
`$ git pr show -t 15143`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15143.diff">https://git.openjdk.org/jdk/pull/15143.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15143#issuecomment-1665581450)

AlanBateman commented on src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:
```diff
@@ -650,6 +650,21 @@ void cleanup(Control[] reqCtls, boolean notifyParent) {
+                            sock.close();
+                            unpauseReader();
+                            if (debug) {
+                                if (sock.isClosed()) {
```
How is it possible for sock.close() to succeed but isClosed to return false?

weibxiao replied:
It is an unreachable code and got removed.

msheppar commented:
why not restructure the finally block a little ... refactor extract methods
flushOutputStream, 
closeConnectionSocket 
... should the outputStream be closed in this finally block, also?

                } finally {
                    flushOutputStream(outStream);
                    closeConnectionSocket(sock);
                    try {
                        unpauseReader();
                    } catch (IOException ie) {
                        if (debug)
                            System.err.println("Connection.cleanup: problem with unpuaseReader " + ie);
                    }
  ...


    private void flushOutputStream (OutputStream outputStream) {
        try {
            outStream.flush();
            outStream.close();
        } catch (IOException ioEx) {
            if (debug)
                System.err.println("Connection.flushOutputStream: OutputStream flush or close problem " + ioEx);
        }
    }
    
    private void closeConnectionSocket (Socket sock) {
        //bug 8313657, socket not closed util GC running
        try {
            sock.close();
        } catch (IOException ioEx) {
            if (debug)
                System.err.println("ConnectioncloseConnectioSocket: problem closing socket: " + ioEx);
        }
    }

AlekseiEfimov commented on src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:
```diff
@@ -650,6 +650,19 @@ void cleanup(Control[] reqCtls, boolean notifyParent) {
                     } catch (IOException ie) {
                         if (debug)
                             System.err.println("Connection: problem closing socket: " + ie);
```
How about combining the debug statement here with the one below and just print the socket isClosed state, something like
```
if (debug) {
    System.err.println("Connection: problem cleaning-up the connection: " + ie);
    System.err.println("Socket isClosed: " + sock.isClosed());
}
```
The text was changed here since the exception can also be thrown by `unpauseReader`

AlekseiEfimov commented on src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:
```diff
@@ -650,6 +650,19 @@ void cleanup(Control[] reqCtls, boolean notifyParent) {
+                            }
+                        } catch (IOException ioe) {
+                            if (debug)
+                                System.err.println("Connection::cleanup problem: " + ioe);
```
Maybe change the message here to highlight the fact that it's the 2nd attemp to clean-up the connection

weibxiao commented:
1) Extracted the methods for finally black of Connection::cleanUP
2) Move the test class to LDAP test folder and updated the test comments.
3) Removed the redundant code and update the import statement.

vyommani commented on src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:
```diff
@@ -684,6 +682,43 @@ void cleanup(Control[] reqCtls, boolean notifyParent) {
     }
 
+    // flush and close output stream
+    private void flushCloseOutputStream() {
```
'flushAndCloseOutputStream' will be more meaningful name.

vyommani commented on src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:
```diff
@@ -684,6 +682,43 @@ void cleanup(Control[] reqCtls, boolean notifyParent) {
+    }
+
+    // close socket
+    private void closeConnectionSocket() {
```
'closeSocketConnection' will be more meaningful.

msheppar replied:
the suggested name was to convey the closing of the  Connection's socket, not a socket connection per se.

msheppar commented:
I ran the test multiple times in mach5 and it executes fine. However, while examining the log output I noticed the "The socket was not closed. " output. This is a little confusing. While looking at the test and exploring the origins of this outout, we see that the test is essentially an othervm execution. As such, I restructured the test to othervm and it executes with some informative output and eliminates the discombobulating output.

I have left a comment in the JBS bug item with some further details and changes for othervm. This gives some imformative output as the test executes.

The origins of the output is the main try block executing in the agentvm and the throwing of a ClassNotFoundException.

If you don't wish to do otherrvm and to avoid the  "The socket was not closed. " output, and if wish to retain the same structure, then consider adding a run command to the test with an arg, and then a conditional on the core test logic e.g.

* @run main SocketCloseTest launcher

and in the main method

    public static void main(String[] args) throws Exception {
        Hashtable<String, Object> props = new Hashtable<>();

if (args.length == 0)  { // only executed in the launched JVM
        props.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
        props.put(Context.PROVIDER_URL, "ldap://localhost:1389/o=example");
        props.put("java.naming.ldap.factory.socket", CustomSocketFactory.class.getName());
        try {
            final DirContext ctx = new InitialDirContext(props);
        } catch (Exception e) {
            if (CustomSocketFactory.customSocket.closeMethodCalledCount() > 0) {
                System.out.println(SOCKET_CLOSED_MSG);
            } else {
                System.out.println(SOCKET_NOT_CLOSED_MSG);
            }
        }
}

        OutputAnalyzer outputAnalyzer = ProcessTools.executeTestJvm("SocketCloseTest_II");
        outputAnalyzer.stdoutShouldContain(SOCKET_CLOSED_MSG);
        outputAnalyzer.stdoutShouldNotContain(SOCKET_NOT_CLOSED_MSG);
        outputAnalyzer.stdoutShouldContain(BAD_FLUSH);
    }

msheppar commented on test/jdk/com/sun/jndi/ldap/SocketCloseTest.java:
```diff
@@ -0,0 +1,170 @@
+        props.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
+        props.put(Context.PROVIDER_URL, "ldap://localhost:1389/o=example");
+        props.put("java.naming.ldap.factory.socket", CustomSocketFactory.class.getName());
+        try {
```
this will execute twice: once in the agentvm main and also, as desired, in the test JVM main.
The agentvm main sees a ClassNotFoundException being thrown and then the output of "The socket was not closed. " in the log c.f. other comments and bug comments for further details to avoid execution in agentvm or use of othervm test mode execution

weibxiao commented:
1) Update the name of the methods
2) Update the test file.

vyommani commented on test/jdk/com/sun/jndi/ldap/SocketCloseTest.java:
```diff
@@ -0,0 +1,172 @@
+    public static class LdapInputStream extends InputStream {
+        private LdapOutputStream los;
+        private ByteArrayInputStream bos;
+        int pos = 0;
```
variable 'pos' is not used can be removed.

vyommani commented on test/jdk/com/sun/jndi/ldap/SocketCloseTest.java:
```diff
@@ -0,0 +1,172 @@
+        @Override
+        public int read() throws IOException {
+            bos = new ByteArrayInputStream(BIND_RESPONSE);
+            int next = bos.read();
```
can be simplified as follows return bos.read();

vyommani commented on test/jdk/com/sun/jndi/ldap/SocketCloseTest.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    public static class CustomSocketFactory extends SocketFactory {
+        public static CustomSocket customSocket = new CustomSocket();
```
you can make 'customSocket' to private.

weibxiao commented:
1) Cannot make CustomSocket be private. The test failed. I believe JDK is using the reflection to load the class.
2) The rest of the changes are based on the comments.

vyommani reviewed (approved):
Latest code looks OK to me.

AlekseiEfimov commented on test/jdk/com/sun/jndi/ldap/SocketCloseTest.java:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    private static class LdapInputStream extends InputStream {
+        private LdapOutputStream los;
```
The `LdapOutputStream` reference is not used by `LdapInputStream`, therefore the `los` field can be removed:
```
private static class LdapInputStream extends InputStream {
-        private LdapOutputStream los;
         private ByteArrayInputStream bos;
 
-        public LdapInputStream(LdapOutputStream los) {
-            this.los = los;
+        public LdapInputStream() {
         }
 
         @Override
@@ -144,7 +142,7 @@ public class SocketCloseTest {
     private static class CustomSocket extends Socket {
         private int closeMethodCalled = 0;
         private LdapOutputStream output = new LdapOutputStream();
-        private LdapInputStream input = new LdapInputStream(output);
+        private LdapInputStream input = new LdapInputStream();
 
         public void connect(SocketAddress address, int timeout) {
         }
```

weibxiao commented:
Updated LdapInputStream.java and remove the unused variable.

msheppar reviewed (approved):
thanks for incorporating the feedback
LGTM

weibxiao commented:
/integrate

AlekseiEfimov commented:
/sponsor
